### DISCUSSION TITLE
image save function execute efficiency question

### DISCUSSION THREAD

**Discussion by zw20045500** [libvips/libvips#2400] [Help wanted]
Hello,
    I found use vips_image_write_to_buffer or vips_pngsave_buffer to change one image format to another format(e.g.  jpg to png),  the tps is low. Looking forward your suggestion.

```C
JNIEXPORT jbyteArray JNICALL
Java_com_criteo_vips_VipsImage_writePNGToArrayNative(JNIEnv *env, jobject obj, jint compression, jboolean palette) {
    jbyteArray ret;
    void *buffer = NULL;
    VipsImage *im = (VipsImage *) (*env)->GetLongField(env, obj, handle_fid);
    size_t result_length = 0;

    if (vips_pngsave_buffer(im, &buffer, &result_length,
                            "compression", compression,
                            "palette", palette)) {
        throwVipsException(env, "Unable to write PNG image buffer");
        return ret;
    }
    ret = (*env)->NewByteArray(env, result_length);
    (*env)->SetByteArrayRegion(env, ret, 0, result_length * sizeof(jbyte), buffer);
    g_free(buffer);
    return ret;
}
```

**Comment by jcupitt** [comment]
You'd need to include more context. libvips is a lazy library, so almost no work happens until the final save. Where does the input `VipsImage` object come from? What work does it represent?

If this is just load + save, there's not much you can do, since almost all processing time will be in the file format libraries, which are usually single-threaded, and which libvips has almost no control over. 

All you can do is change the write parameters. For example, PNG write is very slow and 90% of time is spent in deflate compression. Turn the compression factor down (or pick a different format) and you'll see a good speedup:

```
$ time vips copy wtc.png x.png
real	0m30.667s
user	0m32.647s
sys	0m0.429s
$ time vips copy wtc.png x.png[compression=0]

real	0m7.594s
user	0m7.867s
sys	0m0.564s
$ time vips copy wtc.png x.jpg

real	0m3.843s
user	0m4.547s
sys	0m0.262s
```

**Reply by zw20045500** [reply]
VipsImage is defined for java to call by JNI. In VipsImage(C language) can call libvips.  What's mean  the file format libraries are usually single-threaded. 
     if send many request to process image,  the file format libraries will synchronize to do these request?

**Reply by jcupitt** [reply]
> VipsImage is defined for java to call by JNI. In VipsImage(C language) can call libvips. 

When the save executes, it will force the execution of the code inside the VipsImage as well. So the speed of the save depends on how you made that.

How did you make the VipsImage?

>  What's mean the file format libraries are usually single-threaded.
> if send many request to process image, the file format libraries will synchronize to do these request?

Image compression (for example, to PNG) will only use one CPU, so it can quickly become the bottleneck.

**Reply by zw20045500** [reply]
write png code as follow, call vips_pngsave_buffer(libvips) to save :

```C
JNIEXPORT jbyteArray JNICALL Java_com_criteo_vips_VipsImage_writePNGToArrayNative2
        (JNIEnv *env, jobject obj, jint compression) {
    jbyteArray ret;
    void *buffer = NULL;
    VipsImage *im = (VipsImage *) (*env)->GetLongField(env, obj, handle_fid);
    size_t result_length = 0;

    if (vips_pngsave_buffer(im, &buffer, &result_length, "compression", compression, NULL)) {
        throwVipsException(env, "Unable to write PNG image buffer");
        return ret;
    }
    ret = (*env)->NewByteArray(env, result_length);
    (*env)->SetByteArrayRegion(env, ret, 0, result_length * sizeof(jbyte), buffer);
    g_free(buffer);
    return ret;
}
```

**Reply by jcupitt** [reply]
Yes, but how was the `VipsImage` made?

```C
VipsImage *im = (VipsImage *) (*env)->GetLongField(env, obj, handle_fid);
```

You need to show the code that created that pointer.

**Reply by zw20045500** [reply]
This is JNI function, JNI will make the VipsImage. Now we test vips_pngsave_buffer fucntion is slow, not create JNI create VipsImage

**Reply by jcupitt** [reply]
Yes, but what vips calls did you use to make the `VipsImage`?

As I said, libvips is a lazy library. When you make a series of calls, nothing happens until the final one. For example, if you do this:

1. `x = vips_image_new_from_file("x.tif")`
2. `vips_image_write_to_file(x, "y.png")`

The PNG save will be relatively fast. But if you do this:

1. `x = vips_image_new_from_file("x.tif")`
2. `x = x.gaussblur(100)`
3. `vips_image_write_to_file(x, "y.png")`

Now the save will be much slower, since the blur will be computed inside the PNG save.

**Reply by zw20045500** [reply]
**Yes, but what vips calls did you use to make the VipsImage?**
---- for example, I want make a image to add shadow,  rotate e.g.  a series of operations. So i have to get VipsImage handler to do what i want.

**-----When you make a series of calls, nothing happens until the final one**
thank you. I got it
